### PR TITLE
feat(components): enhance Modal and migrate to TypeScript

### PR DESCRIPTION
**Pull request by connor-baer**
Supercedes #415, #483.

## Approach and changes

- migrate Modal and subcomponents to TypeScript
- add `useModal` hook
- make `ModalFooter` sticky on mobile for when the content exceeds the modal height

## Definition of done

* [x] Development completed
* [x] Reviewers assigned
* [x] Unit and integration tests
* [x] Meets minimum browser support
* [x] Meets accessibility requirements

### CODE REVIEW

File: src/components/Modal/Modal.spec.tsx [hunk 1]
```diff
@@ -15,18 +15,21 @@
 
 import React from 'react';
 
-import { ModalConsumer, ModalProvider } from '.';
+import { render, act, userEvent, wait } from '../../util/test-utils';
```
Comment (fernandofleury) [inline src/components/Modal/Modal.spec.tsx:18]: Why is it named `userEvent`? Aren't we using rtl for this?

Comment (connor-baer) [reply src/components/Modal/Modal.spec.tsx:18]: `@testing-library/user-event` is part of the RTL ecosystem:

> `user-event` is a companion library for React Testing Library that provides more advanced simulation of browser interactions than the built-in `fireEvent` method.

From https://testing-library.com/docs/ecosystem-user-event

File: src/components/Modal/Modal.spec.tsx [hunk 2]
```diff
@@ -15,18 +15,21 @@
...
 import Button from '../Button';
+import { ModalProps } from './Modal';
+import { ModalConsumer, ModalProvider } from './ModalContext';
 
 import * as MockedModal from './Modal';
 
 describe('Modal', () => {
   beforeEach(() => {
-    MockedModal.TRANSITION_DURATION = 0;
+    (MockedModal as any).TRANSITION_DURATION = 0;
```
[the remark this answers is not in the document]
Comment (connor-baer) [reply src/components/Modal/Modal.spec.tsx:28]: Neat! Unfortunately, I get another type error then. I'd rather just leave it.

File: src/components/Modal/Modal.spec.tsx (hunk 1 above)
Comment (fernandofleury) [reply src/components/Modal/Modal.spec.tsx:18]: I just saw from this blogpost: https://kentcdodds.com/blog/common-mistakes-with-react-testing-library. Looks like they are adding lot's of useful goodies (like find*). Good to know!

**Pull request merged**